File: my_py_pkg/my_py_pkg/number_counter.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node

from example_interfaces.msg import Int64
from example_interfaces.srv import SetBool


class NumberCounterNode(Node): 
    def __init__(self):
        super().__init__("number_counter") 
        self.msg_counter_ = 0
        self.subscriber_ = self.create_subscription(
            Int64, "/number", self.callback_number_counter, 10)
        self.publisher_ = self.create_publisher(Int64, "number_count", 10)
        self.reset_counter_servies_ = self.create_service(SetBool, "reset_counter", self.callback_reset_counter)
        self.get_logger().info("Number Count&Publish has been started.")
        

    def callback_number_counter(self, msg):
        self.msg_counter_ += msg.data
        msg.data = self.msg_counter_
        self.publisher_.publish(msg)
        self.get_logger().info(str(msg.data))

    
    def callback_reset_counter(self, request, response):
        if request.data:
            self.msg_counter_ = 0
            response.success = True
            response.message = "Counter has been reset"
        else:
            response.success = False
            response.message = "counter has not been reset"
        return response


def main(args=None):
    rclpy.init(args=args)
    node = NumberCounterNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/number_counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "example_interfaces/msg/int64.hpp"
#include "example_interfaces/srv/set_bool.hpp"

class NumberCounterNode : public rclcpp::Node 
{
public:
    NumberCounterNode() : Node("number_counter") 
    {
        subscriber_ = this->create_subscription<example_interfaces::msg::Int64>("number", 10, 
            std::bind(&NumberCounterNode::callbackNumberCounter, this, std::placeholders::_1));
        publisher_ = this->create_publisher<example_interfaces::msg::Int64>("number_count", 10);
        server_ = this->create_service<example_interfaces::srv::SetBool>(
            "reset_counter", 
            std::bind(&NumberCounterNode::callbackSetBool, this, std::placeholders::_1, std::placeholders::_2));
        RCLCPP_INFO(this->get_logger(), "Number Counter has been started.");
                                
    }

private:

    void callbackSetBool(const example_interfaces::srv::SetBool::Request::SharedPtr request,
                            const example_interfaces::srv::SetBool::Response::SharedPtr response)
    {
        if (request->data == true)
        {
            this->counter_ = 0;
            response->success = true;
            response->message = "Counter has been reset";
        }
        else
        {
            response->success = false;
            response->message = "Counter has not been reset";
        }
    }

    void callbackNumberCounter(const example_interfaces::msg::Int64::SharedPtr msg)
    {
        auto publish_msg = example_interfaces::msg::Int64();
        this->counter_ += msg->data;
        publish_msg.data = this->counter_;
        publisher_->publish(publish_msg);
        RCLCPP_INFO(this->get_logger(), "%ld", publish_msg.data);
    }

    rclcpp::Subscription<example_interfaces::msg::Int64>::SharedPtr subscriber_;
    rclcpp::Publisher<example_interfaces::msg::Int64>::SharedPtr publisher_;
    rclcpp::Service<example_interfaces::srv::SetBool>::SharedPtr server_;
    int counter_ = 0;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<NumberCounterNode>(); 
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}